File: TerminalVoiceOverlay-Windows/Services/DiagLog.cs
using System;
using System.IO;
using System.Text;

namespace TerminalVoiceOverlay.Services
{
    /// <summary>
    /// Minimaler, fehlertoleranter JSON-Lines-Datei-Logger fuer Laufzeit-Diagnose
    /// (Observability-First). Die fensterlose self-contained single-file EXE hat KEINE
    /// sichtbare Konsole — Console.WriteLine ist dort unsichtbar. Dieser Logger schreibt
    /// strukturierte Zeilen in eine feste Datei, die per
    ///   Get-Content "$env:LOCALAPPDATA\TerminalVoiceOverlay\diag.log" -Wait -Tail 30
    /// live mitgelesen werden kann.
    ///
    /// Format: {"ts":"2026-...","ctx":"UIA","msg":"...", &lt;weitere k/v&gt;}
    /// Grundsatz: Logging darf NIE den Ablauf stoeren — jede Methode ist werf-frei.
    /// Single-File-sicher: nutzt LocalApplicationData, NICHT Assembly.Location (IL3000).
    /// </summary>
    public static class DiagLog
    {
        private static readonly object _lock = new();
        private static readonly string _path = ResolvePath();
        private static bool _announced;

        public static string FilePath => _path;

        private static string ResolvePath()
        {
            try
            {
                string dir = Path.Combine(
                    Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                    "TerminalVoiceOverlay");
                Directory.CreateDirectory(dir);
                return Path.Combine(dir, "diag.log");
            }
            catch
            {
                return Path.Combine(Path.GetTempPath(), "TerminalVoiceOverlay-diag.log");
            }
        }

        /// <summary>Schreibt eine JSON-Zeile. Niemals werfend.</summary>
        public static void Write(string ctx, string msg, params (string Key, object? Value)[] extra)
        {
            try
            {
                var sb = new StringBuilder(256);
                sb.Append('{');
                AppendKv(sb, "ts", DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ss.fff"));
                sb.Append(','); AppendKv(sb, "ctx", ctx);
                sb.Append(','); AppendKv(sb, "msg", msg);
                if (extra != null)
                    foreach (var (k, v) in extra)
                    {
                        sb.Append(',');
                        AppendKv(sb, k, v?.ToString() ?? "null");
                    }
                sb.Append('}');

                lock (_lock)
                {
                    RotateIfBig();
                    if (!_announced)
                    {
                        _announced = true;
                        File.AppendAllText(_path,
                            "{\"ts\":\"" + DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ss.fff") +
                            "\",\"ctx\":\"DiagLog\",\"msg\":\"=== session start ===\"}\n",
                            Encoding.UTF8);
                    }
                    File.AppendAllText(_path, sb.ToString() + "\n", Encoding.UTF8);
                }
            }
            catch { /* Logging darf nie crashen */ }
        }

        private static void RotateIfBig()
        {
            try
            {
                var fi = new FileInfo(_path);
                if (fi.Exists && fi.Length > 1_000_000) // 1 MB -> rotieren
                {
                    string bak = _path + ".1";
                    if (File.Exists(bak)) File.Delete(bak);
                    File.Move(_path, bak);
                }
            }
            catch { /* tolerant */ }
        }

        private static void AppendKv(StringBuilder sb, string key, string val)
        {
            sb.Append('"').Append(Escape(key)).Append("\":\"").Append(Escape(val)).Append('"');
        }

        private static string Escape(string s)
        {
            if (string.IsNullOrEmpty(s)) return "";
            return s.Replace("\\", "\\\\").Replace("\"", "\\\"")
                    .Replace("\r", " ").Replace("\n", " ").Replace("\t", " ");
        }
    }
}

File: TerminalVoiceOverlay-macOS/TerminalVoiceOverlay/DiagLog.swift
import Foundation

/// Minimaler, fehlertoleranter JSON-Lines-Datei-Logger fuer Laufzeit-Diagnose
/// (Observability-First). 1:1-Portierung von
/// TerminalVoiceOverlay-Windows/Services/DiagLog.cs.
///
/// Grund fuer die eigene Datei: eine gestartete .app hat KEINE sichtbare Konsole.
/// `NSLog` landet zwar im Unified Log, ist dort aber zwischen Systemmeldungen kaum
/// zu finden und laesst sich nicht einfach mitlesen. Diese Datei dagegen live:
///
///     tail -f ~/Library/Application\ Support/TerminalVoiceOverlay/diag.log
///
/// Format: {"ts":"2026-...","ctx":"Groq","msg":"...", <weitere k/v>}
///
/// Grundsatz wie in Windows: **Logging darf NIE den Ablauf stoeren** — jede Methode
/// schluckt ihre Fehler. Geschrieben wird auf einer eigenen seriellen Queue, damit
/// weder der Audio- noch der UI-Thread auf die Platte wartet.
enum DiagLog {

    /// Ab dieser Groesse wird rotiert (Windows: 1_000_000).
    private static let maxFileBytes: UInt64 = 1_000_000

    private static let queue = DispatchQueue(label: "tvo.diaglog")
    private static var announced = false

    /// Pfad der Log-Datei. Faellt bei Problemen auf das Temp-Verzeichnis zurueck.
    static let filePath: String = resolvePath()

    private static func resolvePath() -> String {
        let fm = FileManager.default
        if let support = fm.urls(for: .applicationSupportDirectory, in: .userDomainMask).first {
            let dir = support.appendingPathComponent("TerminalVoiceOverlay")
            do {
                try fm.createDirectory(at: dir, withIntermediateDirectories: true)
                return dir.appendingPathComponent("diag.log").path
            } catch {
                // faellt unten auf Temp zurueck
            }
        }
        return (NSTemporaryDirectory() as NSString)
            .appendingPathComponent("TerminalVoiceOverlay-diag.log")
    }

    // MARK: - Oeffentliche API (Namen 1:1 wie Windows)

    static func info(_ ctx: String, _ msg: String, _ extra: [(String, Any?)] = []) {
        write(ctx, msg, extra)
    }

    static func warn(_ ctx: String, _ msg: String, _ extra: [(String, Any?)] = []) {
        write(ctx, msg, [("level", "WARN")] + extra)
    }

    static func error(_ ctx: String, _ msg: String, _ err: Error, _ extra: [(String, Any?)] = []) {
        write(ctx, msg, [("level", "ERROR"),
                         ("err", err.localizedDescription),
                         ("type", String(describing: type(of: err)))] + extra)
    }

    /// Laufzeit-Messung. `since` ist ein Startzeitpunkt aus `DiagLog.now()`.
    static func perf(_ ctx: String, _ step: String, since start: CFAbsoluteTime,
                     _ extra: [(String, Any?)] = []) {
        let ms = Int((CFAbsoluteTimeGetCurrent() - start) * 1000)
        write(ctx, "perf", [("step", step), ("ms", ms)] + extra)
    }

    /// Startzeitpunkt fuer `perf` (Windows-Pendant: `Stopwatch.StartNew()`).
    static func now() -> CFAbsoluteTime { CFAbsoluteTimeGetCurrent() }

    /// Schreibt eine JSON-Zeile. Niemals werfend.
    static func write(_ ctx: String, _ msg: String, _ extra: [(String, Any?)] = []) {
        let stamp = timestamp()
        var line = "{"
        line += kv("ts", stamp)
        line += "," + kv("ctx", ctx)
        line += "," + kv("msg", msg)
        for (key, value) in extra {
            let text: String
            if let value = value { text = String(describing: value) } else { text = "null" }
            line += "," + kv(key, text)
        }
        line += "}"

        queue.async {
            if !announced {
                announced = true
                append("{\"ts\":\"\(stamp)\",\"ctx\":\"DiagLog\",\"msg\":\"=== session start ===\"}")
            }
            append(line)
        }
    }

    // MARK: - Intern

    private static func append(_ line: String) {
        rotateIfBig()
        guard let data = (line + "\n").data(using: .utf8) else { return }
        let fm = FileManager.default
        if let handle = FileHandle(forWritingAtPath: filePath) {
            defer { try? handle.close() }
            do {
                try handle.seekToEnd()
                try handle.write(contentsOf: data)
            } catch {
                // Logging darf nie stoeren
            }
        } else {
            fm.createFile(atPath: filePath, contents: data)
        }
    }

    private static func rotateIfBig() {
        let fm = FileManager.default
        guard let attrs = try? fm.attributesOfItem(atPath: filePath),
              let size = attrs[.size] as? UInt64, size > maxFileBytes else { return }
        let backup = filePath + ".1"
        try? fm.removeItem(atPath: backup)
        try? fm.moveItem(atPath: filePath, toPath: backup)
    }

    private static func timestamp() -> String {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS"
        return f.string(from: Date())
    }

    private static func kv(_ key: String, _ value: String) -> String {
        "\"\(escape(key))\":\"\(escape(value))\""
    }

    private static func escape(_ s: String) -> String {
        s.replacingOccurrences(of: "\\", with: "\\\\")
         .replacingOccurrences(of: "\"", with: "\\\"")
         .replacingOccurrences(of: "\r", with: " ")
         .replacingOccurrences(of: "\n", with: " ")
         .replacingOccurrences(of: "\t", with: " ")
    }
}
